### PR TITLE
[WIP] Custom User registration page

### PR DESCRIPTION
**Pull request by zkraime**
<h1>Purpose: </h1> Miscellaneous fixes and additions

<h2> To do </h2>  
- [x] Add basic registration fields (email, first/last name, password) 
- [x] Add FilterSelectMultiple widget for displaying/selecting/assigning Permissions (and Groups?)
- [x] Add email invitation upon new registration on custom page

### CODE REVIEW

**Comment by zkraime** [comment]
@cwisecarver Please see this, thank you

File: admin/common_auth/views.py [hunk 1]
```diff
@@ -24,3 +25,31 @@ def login(request):
...
+    if request.user.is_authenticated():
+        return redirect('auth:login')
+    if request.method == 'POST':
+        form = CustomUserRegistrationForm(request.POST)
+        if form.is_valid():
+            email = form.cleaned_data['email']
+            first_name = form.cleaned_data['first_name']
+            last_name = form.cleaned_data['last_name']
+            password = form.cleaned_data['password']
+            user = MyUser.objects.create_user(username=email,
```
Comment (cwisecarver) [inline admin/common_auth/views.py]: You `MyUserManager.create_user` doesn't take those kwargs. You could change it to accept them, depending on other usages, or you could do this:
```python
        if form.is_valid():
            email = form.cleaned_data['email']
            first_name = form.cleaned_data['first_name']
            last_name = form.cleaned_data['last_name']
            password = form.cleaned_data['password1']
            user = MyUser.objects.create_user(email=email, password=password)
            user.first_name = first_name
            user.last_name = last_name
            user.save()
```

File: admin/templates/register.html [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+          <label for="first_name">First name:</label>
+          <p>{{ form.first_name }}</p>
+          {% if form.first_name.errors %}<span class="text-danger"> {{ form.first_name.errors }}</span>{% endif %}
+      </div>
+      <div class="form-group">
+          <label for="last_name">Last name:</label>
+          <p>{{ form.last_name }}</p>
+          {% if form.last_name.errors %}<span class="text-danger"> {{ form.last_name.errors }}</span>{% endif %}
+      </div>
+       <div class="form-group">
```
Comment (cwisecarver) [inline admin/templates/register.html]: IIRC we discussed this. You probably just haven't gotten around to changing it yet but my version looks like this. You want to have both password fields there so people don't accidentally mistype their password and we don't validate it against the other entry.

```html
          <div class="form-group">
              <label for="password1">Password:</label>
              <p>{{ form.password1 }}</p>
              {% if form.password1.errors %}<span class="text-danger"> {{ form.password1.errors }}</span>{% endif %}
          </div>
       <div class="form-group">
          <label for="password2">Confirm Password:</label>
          <p>{{ form.password2 }}</p>
          {% if form.password2.errors %}<span class="text-danger"> {{ form.password2.errors }}</span>{% endif %}
      </div>
```

File: admin/templates/register.html [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+        <script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
+    <![endif]-->
+  </head>
+ <body class="register-page">
+  <div class="register-box">
+    <div class="register-logo">
+      <b>OSF</b>Admin
+    </div>
+    <div class="register-box-body">
+      <p class="login-box-msg">Register a new account</p>
```
Comment (cwisecarver) [inline admin/templates/register.html]: In order to see if there were any non-field-specific errors I added this:
```html
      <div>
          {{ form.errors }}
      </div>
```

You could make it much prettier and make it only show non-field errors. This just got the job done for debugging. Here's some documentation on it. https://docs.djangoproject.com/en/1.9/topics/forms/#rendering-form-error-messages

File: admin/templates/register.html [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+      <a href="{% url 'auth:login' %}" class="text-center">I already have an account</a>
+    </div><!-- /.form-box -->
+  </div><!-- /.register-box -->
+
+   <!-- jQuery 2.1.4 -->
+  <script src="{% static 'js/jquery.min.js' %}"></script>
+  <!-- Bootstrap 3.3.2 JS -->
+  {% render_bundle 'vendor' %}
+  <!-- AdminLTE App -->
+  {% render_bundle 'admin-base-page' %}
```
Comment (cwisecarver) [inline admin/templates/register.html]: I had to do this for the last PR too and I realized I didn't note it. This bundle doesn't exist anymore. By changing admin-base-page to app.min I got pages to load again. Nothing like a tightly-coupled asset framework to give you a hard 500 error if your assets aren't in the right place.

File: admin/common_auth/forms.py [hunk 5]
```diff
@@ -15,3 +16,11 @@ def __init__(self, *args, **kwargs):
         super(CustomUserRegistrationForm, self).__init__(*args, **kwargs)
         self.fields['first_name'].required = True
         self.fields['last_name'].required = True
+
```
Comment (cwisecarver) [inline admin/common_auth/forms.py]: I had to change the fields list on this class to 
```python
fields = ['password1', 'password2', 'first_name', 'last_name', 'email', 'is_active', 'is_staff',
            'is_superuser', 'groups', 'user_permissions', 'last_login', ]
```

File: admin/common_auth/views.py [hunk 6]
```diff
@@ -24,3 +25,31 @@ def login(request):
...
+    # Creates User, takes selected permissions and assigns it to user, then saves resetform for email invitation
+    if request.user.is_authenticated():
+        return redirect('auth:login')
+    if request.method == 'POST':
+        form = CustomUserRegistrationForm(request.POST)
+        if form.is_valid():
+            email = form.cleaned_data['email']
+            first_name = form.cleaned_data['first_name']
+            last_name = form.cleaned_data['last_name']
+            password = form.cleaned_data['password']
```
Comment (cwisecarver) [inline admin/common_auth/views.py]: After changing the field list on the form I had to change this line to
```python
            password = form.cleaned_data['password1']

```

File: admin/templates/register.html [hunk 7]
```diff
@@ -0,0 +1,73 @@
+{% load render_bundle from webpack_loader %}
```
Comment (cwisecarver) [inline admin/templates/register.html]: Probably not part of this PR but both this and the login template could inherit from base.html. If you move the navigation you want hidden in base.html into a block and just pass an empty navigation block from the child templates where you want it hidden. This would give you more reuse and less duplication.

File: admin/common_auth/views.py [hunk 8]
```diff
@@ -24,3 +25,31 @@ def login(request):
 def logout(request):
     logout_user(request)
     return redirect('auth:login')
+
+def register(request):
+    # Creates User, takes selected permissions and assigns it to user, then saves resetform for email invitation
+    if request.user.is_authenticated():
+        return redirect('auth:login')
```
Comment (cwisecarver) [inline admin/common_auth/views.py]: Might want to put a flash message call here to tell the user why they can't register an account before redirecting them just for user-friendliness.

File: admin/templates/registration/password_change_done.html [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+<div class="breadcrumbs">
+<a href="{% url 'admin:index' %}">{% trans 'Home' %}</a>
+&rsaquo; {% trans 'Password change' %}
+</div>
+{% endblock %}
+
+{% block title %}{{ title }}{% endblock %}
+{% block content_title %}<h1>{{ title }}</h1>{% endblock %}
+{% block content %}
+<p>{% trans 'Your password was changed.' %}</p>
```
Comment (cwisecarver) [inline admin/templates/registration/password_change_done.html]: Getting setup for translating is a good idea but there are a lot of places throughout this PR where we aren't doing it. Not sure what the requirement for i18n was for the story but my vote would be do it everywhere we have strings or don't do it now. It'll be a lot easier to add it in later if it's one or the other.

**Comment by cwisecarver** [comment]
@Ghalko, @zkraime I'm done. Let me know if there are questions or comments. If I saw the same issues that I commented on in the last PR I tried to leave them out of here. There are some duplicated things though like all the email stuff. I'm sure it was needed in both in order to accomplish both. Might be worth merging these together and then sending them on to @sloria  as one when the time comes just to make sure we're not undoing changes done in the prior PR.

**Comment by zkraime** [comment]
@cwisecarver 
Thank you for the review! 
I apologize - on Friday when I had left to go home, I had not quite finished the update PR and was going to before long, but then vacation plans overtook. 
I will be able to get to it and finish by Wednesday. 
Thanks again

**Comment by cwisecarver** [comment]
No need to apologize @zkraime. It's the holidays. I'm not sure when @Ghalko is going to be back to look at it. I'd work around his schedule.

**Comment by zkraime** [comment]
@cwisecarver Roger that

**Comment by Ghalko** [comment]
@zkraime What is the status of this?

**Comment by zkraime** [comment]
@Ghalko I almost have this done, I plan to finish it and get it to Steve after we meet up tomorrow (hopefully we don't get snowed out)

File: admin/templates/register.html (hunk 3 above)
Comment (zkraime) [inline admin/templates/register.html]: @cwisecarver,
For this, should I just add {{ form.non_field_errors }} at the bottom of the form? 
Do you think it's important to style it?

File: admin/templates/registration/password_change_done.html (hunk 9 above)
Comment (zkraime) [inline admin/templates/registration/password_change_done.html]: Ok for now I'll just exclude it

File: admin/templates/register.html (hunk 3 above)
Comment (cwisecarver) [inline admin/templates/register.html]: I think a user would want the errors at the top of the form.

I think you can loop through both form.non_field_errors and form.errors and put it into a list.

```html
{% if form.non_field_errors %}
<ul>
{% for error in form.non_field_errors %}
       <li class="errorclass">{{ error }}</li>
{% endfor %}
</ul>
{% endif %}
```

**Comment by zkraime** [comment]
@cwisecarver This should be good now, but if you can give a quick pass so I can forward to @sloria that'd be great; thank you

**Comment by cwisecarver** [comment]
@zkraime Looks good, onward to @sloria

**Comment by zkraime** [comment]
@sloria I need to update my osf first, for this and for https://github.com/CenterForOpenScience/osf.io/pull/4718
Ill get that done asap
